### DISCUSSION TITLE
get_location more robust

### DISCUSSION THREAD

**Discussion by WolfgangFahl** [zauberzeug/nicegui#4075] [Q&A]
### Question

I tried adapting https://github.com/zauberzeug/nicegui/discussions/257 for http://nicetrack.bitplan.com/map

but get
```python
 File "/home/wf/source/python/nicetrack/.venv/lib/python3.10/site-packages/nicetrack/map.py", line 29, in setup_ui
    location = await self.get_location()
  File "/home/wf/source/python/nicetrack/.venv/lib/python3.10/site-packages/nicetrack/map.py", line 54, in get_location
    response = await ui.run_javascript('''
  File "/home/wf/source/python/nicetrack/.venv/lib/python3.10/site-packages/nicegui/client.py", line 209, in send_and_wait
    return await JavaScriptRequest(request_id, timeout=timeout)
  File "/home/wf/source/python/nicetrack/.venv/lib/python3.10/site-packages/nicegui/javascript_request.py", line 28, in __await__
    raise TimeoutError(f'JavaScript did not respond within {self.timeout:.1f} s') from e
TimeoutError: JavaScript did not respond within 5.0 s
```

may be because of a popup blocker. **How can i make this more robust?**

https://github.com/WolfgangFahl/nicetrack/blob/main/nicetrack/map.py has the code:
```python
async def get_location(self):
        """
        Fetches the current geographic location of the user via JavaScript.
        Returns the coordinates as a tuple (latitude, longitude).
        """
        location=None
        response = await ui.run_javascript('''
            return await new Promise((resolve, reject) => {
                if (!navigator.geolocation) {
                    reject(new Error('Geolocation is not supported by your browser'));
                } else {
                    navigator.geolocation.getCurrentPosition(
                        (position) => {
                            resolve({
                                latitude: position.coords.latitude,
                                longitude: position.coords.longitude,
                            });
                        },
                        () => {
                            reject(new Error('Unable to retrieve your location'));
                        }
                    );
                }
            });
        ''', timeout=5.0)
        if 'latitude' in response:
            location=response['latitude'], response['longitude']
        return location
```

**Comment by falkoschindler** [comment]
Hi @WolfgangFahl,

What do you mean with "more robust"? What is the desired behavior?
Instead of rejecting if there is no `navigator.geolocation`, you could `resolve` with an error string. Then you could check if `response` is a string and show a notification in this case.

**Comment by WolfgangFahl** [comment]
@falkoschindler i would like to translate the javascript error handling to python error handling. Also a retry might be possible or a manual selection of the position. I am not clear on the expected behavior of the geolocation handling in browsers. There seems to be a not of interactive fiddling with settings necessary e.g. my browser will ask for permisson and in addition on macos system permission is also required. A robust solution would work accordingly and just notify that things will only work if settings and permissions have been set. The whole show seem to be quite complicated and it would be nice to have solution that hides this complexity accordingly and is reusable so not every developer has to think about all these details.
